PLEASE MERGE SECOND :: Mutliclass predictions tested and fixed

Aylr opened the pull request:
## Background

- It was discovered that the multiclass predictions were not nearly as robust as they needed to be.

## Minimal script to test

```
import healthcareai

dataframe = healthcareai.load_dermatology()
df = dataframe.copy()

print(dataframe.head(5))

dataframe.drop(['target_num'], axis=1, inplace=True)

classification_trainer = healthcareai.SupervisedModelTrainer(
    dataframe,
    predicted_column='target_str',
    model_type='classification',
    grain_column='PatientID',
    impute=True)

trained_lr = classification_trainer.logistic_regression()

preds = trained_lr.make_predictions(df)
print(preds)
```

mmastand commented:
The following failed:
```
import healthcareai


def main():
    """Template script for using healthcareai to train a classification model."""
    # Load the included dermatology sample data that has 6 classes
    dataframe = healthcareai.load_dermatology()
    dataframe = dataframe.drop(dataframe.index[300:])
    # Peek at the first 5 rows of data
    print(dataframe.head(5))

    # Drop columns that won't help machine learning
    dataframe.drop(['target_num'], axis=1, inplace=True)

    # Step 1: Setup a healthcareai classification trainer. This prepares your data.
    classification_trainer = healthcareai.SupervisedModelTrainer(
        dataframe,
        predicted_column='target_str',
        model_type='classification',
        grain_column='PatientID',
        impute=True)

    # Step 2: train some models and look at their confusion matrices


    # Train a logistic regression model
    trained_rf = classification_trainer.random_forest()

    trained_rf.save()

    # # # # # Predict
    dataframe = healthcareai.load_dermatology()
    dataframe = dataframe.drop(dataframe.index[:301])
    # Peek at the first 5 rows of data


    # Drop columns that won't help machine learning
    dataframe.drop(['target_num'], axis=1, inplace=True)
    dataframe = dataframe.reset_index(drop=True)

    print(dataframe.head(5))

    print(dataframe.shape)
    # predict.
    trained_model = healthcareai.load_saved_model('2017-11-14T07-11-49_classification_RandomForestClassifier.pkl')

    predictions = trained_model.make_predictions(dataframe)
    print(predictions.head())


if __name__ == "__main__":
    main()

```

This code fails when making predictions with the following output:
```
      x32     x33   x34        target_str  
0  Class3  Class3  36.0     lichen planus  
1  Class1  Class2  50.0     lichen planus  
2  Class2  Class0  34.0  pityriasis rosea  
3  Class2  Class0  17.0         psoriasis  
4  Class2  Class0  24.0         psoriasis  

[5 rows x 36 columns]
(65, 36)
Trained model loaded from file: 2017-11-14T07-11-49_classification_RandomForestClassifier.pkl
    Type: <class 'healthcareai.trained_models.trained_supervised_model.TrainedSupervisedModel'>
    Model type: <class 'sklearn.model_selection._search.RandomizedSearchCV'>
/Users/michael.mastanduno/Documents/Repos/healthcareai-py/healthcareai/common/transformers.py:274: FutureWarning: specifying 'categories' or 'ordered' in .astype() is deprecated; pass a CategoricalDtype instead
  categories=self.categorical_levels[col])
Traceback (most recent call last):
  File "/Users/michael.mastanduno/Documents/Repos/healthcareai-py/example_multiclass_2.py", line 40, in <module>
    main()
  File "/Users/michael.mastanduno/Documents/Repos/healthcareai-py/example_multiclass_2.py", line 35, in main
    predictions = trained_model.make_predictions(dataframe)
  File "/Users/michael.mastanduno/Documents/Repos/healthcareai-py/healthcareai/trained_models/trained_supervised_model.py", line 194, in make_predictions
    self.model.classes_,
AttributeError: 'RandomizedSearchCV' object has no attribute 'classes_'
```

Aylr commented:
1. Great idea for the sorted. Captured in #441

Aylr commented:
Addressing comments above:
2. Yes, I have unit tests for:
>Single row prediction works and is the same as multi-row
>Multi row preds with new category or nan work
>Single row preds with new category or nan works

3. The bug you found has been fixed and there are tests in place.

levithatcher commented on healthcareai/common/helpers.py:
```diff
@@ -1,6 +1,8 @@
 import math
 import sklearn
 
```
Wasn't from this change, but would we only want to import the functionality needed (and not the whole package)?

Aylr replied:
https://softwareengineering.stackexchange.com/questions/187403/import-module-vs-from-module-import-function

levithatcher commented on healthcareai/advanced_supvervised_model_trainer.py:
```diff
@@ -555,6 +543,7 @@ def _create_trained_supervised_model(self, algorithm, include_factor_model=True)
             test_set_class_labels=test_set_class_labels,
             test_set_actual=self.y_test,
             metric_by_name=self.metrics(algorithm),
+            class_labels=class_labels,
```
Was this supposed to be `= self.class_labels` (like on 525 above)?

Also, the lack of whitespace around `=` doesn't match other sections...

Aylr replied:
1. Good eye, except that is set elsewhere on 519-525.
2. PEP8 says no spaces around equal signs in function arguments.

mmastand commented:
@Aylr I ran this through my tests and I'm happy with this PR now.

Aylr commented:
@levithatcher regarding your earlier comments:
>Got AttributeError: module 'healthcareai' has no attribute 'load_dermatology'
>After importing the data manually, I saw this:
I could not reproduce either one.

I suspect you have an version of healthcareai installed in your python environment and the `.load_dermatology()` isn't in that version. And, if that's true, unfortunately you likely weren't even testing the new code anyway, since you loaded in the installed version rather than the local module.
